# 4.5 Data Consistency Check

## This script contains the following points:
Note: Corresponding steps for exercise task shown as (Step x)

### 1 Exercise 4.5 Reading and Examples

1.1 Importing Libraries, Defining Project Path, and Importing Datasets
Importing orders_wrangled.csv dataset
Importing products.csv dataset
1.2 Mixed-Type Data Example
1.3 Missing Values
Finding Missing Values in 'df_prods'
Addressing Missing Values in 'df_prods'
1.4 Duplicates
Finding Duplicates in 'df_prods_clean'
Addressing Duplicate Values in 'df_prods_clean'

### 2 Exercise 4.5 Task

(Step 2) Observing df_ords and df_prod with descriptive statistics
(Step 2.a) Addressing suspected typos in product prices
(Step 3) Check for mixed-type data in df_ords dataframe
(Step 4) Fix mixed-type data
(Step 5) Check for missing values in df_ords
(Step 6) Address missing values in df_ords
(Step 7) Check for duplicate values in df_ords
(Step 8) Address duplicate values in df_ords
(Step 9.a) Exporting 'df_ords' (as 'orders_checked.csv')
(Step 9.b) Exporting 'df_prods_clean_no_dups' (as 'products_checked.csv')

In [10]:
#Importing Libraries
import pandas as pd
import numpy as np
import os

In [11]:
path = r'/Users/ronaldcameron/Documents/Data Analytics/Achievement 4 - Python/02:2023 Insta Cart Basket Analysis/Data'

In [12]:
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

In [13]:
df_ords = pd.read_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'), index_col =False)

In [14]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [15]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [16]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [17]:
df_prods.shape

(49693, 5)

In [18]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [22]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
df_prods_clean_no_dups.shape

(49672, 5)

In [24]:
products_checked =df_prods_clean_no_dups

In [33]:
#Export products_checked
products_checked.to_csv (r'/Users/ronaldcameron/Documents/Data Analytics/Achievement 4 - Python/02:2023 Insta Cart Basket Analysis/Data/Prepared Data/products_checked.csv')

Task 4.5


In [39]:
#Q2 Investigate Data
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [37]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [38]:
df_ords.shape

(3421083, 7)

In [46]:
#Q3 Check for mix-type data
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
     print (col)

In [47]:
#Q5 Check for missing values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the 'days_since_prior_order' column

In [49]:
#Q6 Create missing value subset
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() ==True]

In [50]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [51]:
#Create clean subset
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [52]:
df_prods_clean

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49677 rows x 5 columns]

In [53]:
#Q7 Check for duplicates
df_ords_dups = df_ords_clean[df_ords_clean.duplicated()]

In [54]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [55]:
#Q8 -- no duplicates were found

In [56]:
orders_checked = df_ords_clean

In [57]:
#Q9 Exposrt orders_checked
orders_checked.to_csv (r'/Users/ronaldcameron/Documents/Data Analytics/Achievement 4 - Python/02:2023 Insta Cart Basket Analysis/Data/Prepared Data/orders_checked.csv')